Source/OxyPlot/Series/BarSeries/IBarSeries.cs

```diff
@@ -0,0 +1,73 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="IBarSeries.cs" company="OxyPlot">
+//   Copyright (c) 2020 OxyPlot contributors
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace OxyPlot.Series
+{
+    using OxyPlot.Axes;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Defines the functionality of a bar series.
+    /// </summary>
+    public interface IBarSeries
+    {
+        /// <summary>
+        /// Gets the bar width.
+        /// </summary>
+        double BarWidth { get; }
+
+        /// <summary>
+        /// Gets the <see cref="CategoryAxis"/> the bar series uses.
+        /// </summary>
+        CategoryAxis CategoryAxis { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether the bar series is visible.
+        /// </summary>
+        bool IsVisible { get; }
+
+        /// <summary>
+        /// Gets or sets the manager of the bar series.
+        /// </summary>
+        BarSeriesManager Manager { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="PlotModel"/> the bar series belongs to.
+        /// </summary>
+        PlotModel PlotModel { get; }
+
+        /// <summary>
+        /// Gets the <see cref="ValueAxis"/> the bar series uses.
+        /// </summary>
+        Axis ValueAxis { get; }
+
+        /// <summary>
+        /// Updates the valid data.
+        /// </summary>
+        void UpdateValidData();
+
+        // TODO: replace the following members by IReadOnlyList<BarItemBase> once we get rid of .NET 4.0.
```

Comment:
There was a specific request to keep Net Framework 4.0 support quite recently (or atleast it was readded in #1452), so I'm not sure if these comments are useful.

Would it be insane to provide our own type to expose these members so that we don't have the interface splatted into other types', making a future refit painless for everyone (e.g. just change `CustomReadOnlyListWrapperType` with `IReadOnlyList` and update the writes: minimal code change for consumers). This type could even implement `IReadOnlyList` for compilation targets that support it, keeping conditions in one place. Are there other places we could use such a type, or is it just a barmy idea?

Reply:
_(I do not quite agree on supporting .NET 4.0 for much longer and I hate that it forces us to do these kinds of considerations which should be a non-issue nowadays, but that's not the issue here.)_

I think there are quite some places where `IList<T>` is used that should use `IReadOnlyList<T>` instead. In this specific case this was not possible because of lack of covariance.

Possibly slightly insane idea: How about we define `IReadOnlyList<T>` (and maybe also `IReadOnlyCollection<T>`) ourselves (conditionally compiled only for NET40) and a custom list class implementing this (again only for NET40)? 
I'll do some experiments to see if this could even remotely work.

Reply:
I'd be hesitant to use the same names, because the 4.0 library could be used by 4.5 code and that might create confusion, but it would be nice to replace all the horridness with `IList<T>` with some well-typed horridness that can be quickly and easily removed if/when 4.0 support is dropped.

Reply:
Of course we could also use this [package](https://www.nuget.org/packages/ReadOnlyCollectionInterfaces/) ([GitHub link](https://github.com/mausch/ReadOnlyCollections)) for NET4.0. 

You are right, it would not be possible to use the 4.0 library from 4.5 (it would complain when loading the package about ambiguous types I think), but is this a problem?

Reply:
Thanks for the link: I went looking for that project but couldn't find it. That said, I think it's probably too much for our purposes.

It shouldn't be ambiguous so long as the fully qualified name is distinct (which I think was the idea); I just think it could create confusion for users, and that being up-front that this is a thin wrapper over what is usually an array (I think) is a good idea. I suppose we need to decide whether it is important to improve the 4.0 interface, or just make life easier for everybody else. Something as simple as https://gist.github.com/VisualMelon/d8baadc9f28f8d5f1bf60bf9eaf8c318 might suffice.

Reply:
Well this package just contains the three interfaces `IReadOnlyCollection<T>`, `IReadOnlyList<T>` and `IReadOnlyDictionary<T>`, so I wouldn't necessarily say it's too much ;).
However I realized we can't use this anyway, since the assembly is not signed, so doesn't matter anyway. If we would want to go this way, we would have to define the interfaces ourselves (which wouldn't be too difficult).

Your solutions seems neat, especially the implicit cast from `List<T>`. However there is one thing I am a bit concerned about: When we ever finally get rid of .NET 4.0, this will be a breaking change for all interfaces using `ReadOnlyList<T>` (as they will be changed to `IReadOnlyList<T>`). Ok granted, probably this change will not be noticed by most consumers of the API, so maybe this is the best option we have...

Reply:
Yeah, maybe I'm being silly, and we should just provide a conditionally defined `IReadOnlyList<T>` along with a wrapper type to re-expose `IList<T>`.

Reply:
I think this topic warrants its own issue, so I'll open a new one.

Reply:
Good idea, and I don't think it should block this being merged.

Reply:
Well I don't mind leaving this open for a couple more days until we have figured out what to do rather than going back later and adapting this to whatever solution we settle on.